[TACHYON-590] Implement the wildcard support for shell cmds with 1 argument

tianyin opened the pull request:
Per our discussion, we add wildcard support for shell cmds in multiple batches to ease the code review.

This PR contains the implementation of wildcard support for cmds with 1 argument, including the following cmds:
{cat, ls, lsr, rm, rmr, tail, touch, fileinfo, location, report, pin, unpin, free, du}

while the following cmds does not support wildcards by their semantic:
{mkdir, touch, count}
(where wildcard is treated as a regular character)

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4401/
Test PASSed.

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    Assert.assertEquals(expectedOutput, mOutput.toString());
+  }
+  
+  private String getLocationResultStr(String path, int fid, TachyonFS tfs) throws IOException {
```
NIT: What about putting this `private` method at the end of the class to put the `@Test` together?

uronce-cc replied:
`fid` can be gotten once we have `path` and `tfs`?

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar1")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foo/foobar2")).isInMemory());
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/bar/foobar3")).isInMemory());
```
Assert `/testWildCards/foobar4` is not in memory here?

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    TFsShellUtilsTest.resetTachyonFileHierarchy(mTfs, WriteType.MUST_CACHE);
+    
+    mFsShell.run(new String[] {"report", "/testWildCards/*/foo*"});
+    String expectedOutput = "";
```
Better to use `StringBuilder`.

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    expectedOutput += "/testWildCards/foo/foobar1 with file id " 
+        + mTfs.getFileId(new TachyonURI("/testWildCards/foo/foobar1")) 
+        + " has reported been report lost.\n";
+    expectedOutput += "/testWildCards/foo/foobar2 with file id " 
```
These output messages are of the same format, may have a utility method.

uronce-cc replied:
Or we can have a `String[]` containing the paths like `/testWildCards/foo/foobar1, /testWildCards/foo/foobar2....`, then iterate over the array and append output to `expectedOutput`, then we can get rid of the redundant codes.

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    TachyonConf tachyonConf = mLocalTachyonCluster.getMasterTachyonConf();
+    
+    int ret = mFsShell.run(new String[] {"free", "/testWild*/foo/*"});
+    CommonUtils.sleepMs(null, tachyonConf.getInt(Constants.WORKER_TO_MASTER_HEARTBEAT_INTERVAL_MS,
```
The `free` command will block, why we need to sleep here?

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    Assert.assertTrue(mTfs.getFile(new TachyonURI("/testWildCards/foobar4")).isInMemory());
+    
+    ret = mFsShell.run(new String[] {"free", "/testWild*/*/"});
+    CommonUtils.sleepMs(null, tachyonConf.getInt(Constants.WORKER_TO_MASTER_HEARTBEAT_INTERVAL_MS,
```
Why sleep here?

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    TachyonURI expf2 = new TachyonURI("/testWildCards/foo/foobar2");
+    TachyonURI expf3 = new TachyonURI("/testWildCards/bar/foobar3");
+    
+    String expected = "";
```
`StringBuilder`?

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    expected +=
+        String.format(format, FormatUtils.getSizeFromBytes(30),
+            TFsShell.convertMsToDate(mTfs.getFile(expf3).getCreationTimeMs()), "In Memory",
+            expf3.getPath());
```
If we have an array of sizes and an array of paths, then these `exptected += ` can be put into a `for` loop.

uronce-cc commented on shell/src/test/java/tachyon/shell/TFsShellTest.java:
```diff
@@ -658,4 +647,238 @@ public void duTest() throws IOException {
+    
+    
+    TachyonURI expf4 = new TachyonURI("/testWildCards/foobar4");
+    expected += 
```
These `exptected += ` can also be put into a for loop.

uronce-cc commented:
@tianyin Finally, we have `tfs load /data/*` to load all data into memory, I always want this command before. :+1:   Just left a few minor comments for reducing code redundancy.

tianyin commented:
@uronce-cc thanks a lot, Chang! You are always so helpful!

I'll address the comments asap (not in town today...)

I'll remove the single quote. I'm not a fan of it. Just saw in previous code there were some place use it but some place does not. So I just want to make it consistent. 

For the ``sleep`` in ``free``, I followed the ``freeTest`` so perhaps it's not necessary at all...

tianyin commented:
@uronce-cc I have addressed all your comments except the ``sleep`` one. The sleep is needed otherwise the test case cannot pass (as the original ``freeTest``). 

Take a look :-)

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4508/
Test PASSed.

uronce-cc commented:
+1 LGTM

calvinjia commented:
LGTM
